### ISSUE TITLE
compile windows installer against lua 5.4

### ISSUE THREAD

**Issue by prabirshrestha** [vim/vim#6358]
Lua 5.4 is released. It would be good if the windows installer ships with v5.4 instead of v5.3. https://www.lua.org/news.html
I don't see the binaries here yet. http://luabinaries.sourceforge.net/download.html

Here was some of the benchmarks I ran. luajit is the fastest but 5.4 seems to have improved a lot in perf compared to 5.3.

```lua
local a = {}
for i = 1, 1e7 do
        a[#a +1] = i
end
print(#a)
```

lua 5.3
```
$ time lua a.lua
10000000

real    0m2.605s
user    0m2.507s
sys     0m0.093s
```

lua 5.4
```
$ time ~/code/others/lua-5.4.0/install/bin/lua a.lua
10000000

real    0m0.469s
user    0m0.361s
sys     0m0.106s
```

luajit

```
time ~/luajit/bin/luajit a.lua
10000000

real    0m0.120s
user    0m0.056s
sys     0m0.063s
```

vim source code already seems to support 5.4.

```
commit 830e3583da682eee41ed6b027c9912547d4022c0 (tag: v8.1.0305)
Author: Bram Moolenaar <Bram@vim.org>
Date:   Tue Aug 21 14:23:35 2018 +0200

    patch 8.1.0305: missing support for Lua 5.4 32 bits on Unix

    Problem:    Missing support for Lua 5.4 32 bits on Unix.
    Solution:   Define lua_newuserdatauv. (Kazunobu Kuriyama)
```

**Comment by k-takata** [comment]
I think it's better to wait until LuaBinaries is released.

**Comment by jamessan** [comment]
> Lua 5.4 is released. It would be good if the windows installer ships with v5.4 instead of v5.3

Lua's [versioning](https://www.lua.org/versions.html#numbering) isn't like typical versioning.  Each 5.x is a different language, without API or ABI compatibility.

LuaJIT only guarantees compatibility with Lua 5.1, although some work has been done to support features from newer Lua languages.  I would say Lua 5.1 should be the main focus, but Vim currently fails its tests when built against 5.1.

**Comment by k-takata** [comment]
> Lua's [versioning](https://www.lua.org/versions.html#numbering) isn't like typical versioning. Each 5.x is a different language, without API or ABI compatibility.

It is similar to Python's versioning, isn't it?

**Comment by jamessan** [comment]
> It is similar to Python's versioning, isn't it?

No.  Python at least strives to maintain forward API compatibility between X.Y releases, usually only removing things if there have been deprecation cycles.  Since 3.2 there's even an ABI guarantee for a certain surface of the library.

Lua doesn't provide any guarantees between X.Y releases.

**Comment by k-takata** [comment]
From Vim's point of view, it is not true.
You can find several `#if PY_VERSION_HEX >= 0x030X01f0` in `if_python3.c`.

**Comment by jamessan** [comment]
> You can find several `#if PY_VERSION_HEX >= 0x030X01f0` in `if_python3.c`.

Many of which are for _new_ functionality introduced in that version, not backwards incompatible changes.  Looking at Vim's code isn't particularly relevant since provides no insight into how and why Python managed any backwards incompatible changes.

I never stated that backwards incompatible changes don't happen in Python.  Simply that there is an effort to maintain compatibility and to slowly roll out changes that will break compatibility so there is time to adapt.  None of that can be said for Lua, and that's an intentional choice.
